File: practice/mathematics/number_theory/manasa_loves_math/manasa_loves_math.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <string>
#include <map>

using namespace std;

map<int, int> counter(const string & i_s)
{
    map<int, int> c;
    
    string::const_iterator it = i_s.begin();
    for (; it != i_s.end(); ++it)
    {
        c[*it]++;
    }
    
    return c;
}

map<int, int> counter(int i_n)
{
    string s = to_string(i_n);
    return counter(s);
}

bool contains(const map<int, int> & i_cnt1, const map<int, int> & i_cnt2)
{
    map<int, int>::const_iterator it = i_cnt1.begin();
    for (; it != i_cnt1.end(); ++it)
    {
        int key = it->first;
        if (i_cnt2.find(key) == i_cnt2.end())
        {
            return false;
        }
        else if (i_cnt2.at(key) < i_cnt1.at(key))
        {
            return false;
        }
    }
    
    return true;
}

int main()
{
    const vector<int> d8_1{0, 8};
    const vector<int> d8_2{16, 24, 32, 40, 48, 56, 64, 72, 80, 88, 96, 69, 27, 46, 65, 84, 23, 42, 61};
    
    vector<map<int, int>> div_8;
    for (int val = 13; val < 125; ++val)
    {
        div_8.push_back(counter(8 * val));
    }
    
    int t;
    string s;
    
    cin >> t;
    while (t--)
    {
        cin >> s;
        if (s.size() == 1)
        {
            int v = atoi(s.c_str());
            if (find(d8_1.begin(), d8_1.end(), v) != d8_1.end())
            {
                cout << "YES" << endl;
            }
            else
            {
                cout << "NO" << endl;
            }
        }
        else if (s.size() == 2)
        {
            int v = atoi(s.c_str());
            if (find(d8_2.begin(), d8_2.end(), v) != d8_2.end())
            {
                cout << "YES" << endl;
            }
            else
            {
                cout << "NO" << endl;
            }
        }
        else
        {
            map<int, int> c = counter(s);
            bool ans = false;
            
            vector<map<int, int>>::iterator it = div_8.begin();
            for (; it != div_8.end(); ++it)
            {
                if (contains(*it, c))
                {
                    ans = true;
                    break;
                }
            }
            if (ans)
            {
                cout << "YES" << endl;
            }
            else
            {
                cout << "NO" << endl;
            }
        }
    }
    
    return 0;
}

File: practice/mathematics/number_theory/manasa_loves_math/manasa_loves_math.hs
module Main where

import qualified Data.Map as M
import Control.Monad (forM_)
import Data.Maybe

d8_1 :: [Int]
d8_1 = [0, 8]

d8_2 :: [Int]
d8_2 = [16, 24, 32, 40, 48, 56, 64, 72, 80, 88, 96, 69, 27, 46, 65, 84, 23, 42, 61]

div_8 :: [M.Map Char Int]
div_8 = [counter $ show v | v <- [104, 112..992]]

addKey :: Char -> M.Map Char Int -> M.Map Char Int
addKey k m = M.insertWith (+) k 1 m

counter :: String -> M.Map Char Int
counter s = foldl (\acc k -> addKey k acc) M.empty s

contains :: M.Map Char Int -> M.Map Char Int -> Bool
contains c1 c2 = M.foldrWithKey (\k v r -> check k v c2 && r) True c1
    where
        check :: Char -> Int -> M.Map Char Int -> Bool
        check k v c
            | f == Nothing || v > v' = False
            | otherwise              = True
            where
                f  = M.lookup k c
                v' = fromJust f

solve :: String -> String
solve s
    | length s == 1 = if elem (read s :: Int) d8_1 then "YES" else "NO"
    | length s == 2 = if elem (read s :: Int) d8_2 then "YES" else "NO"
    | otherwise     = solve' s
    where
        solve' :: String -> String
        solve' s = if ans then "YES" else "NO"
            where
                c = counter s
                ans = foldr (\c' acc -> contains c' c || acc) False div_8

main :: IO ()
main = do
    t <- readLn :: IO Int
    forM_ [1..t] $ \_ -> do
        s <- getLine
        putStrLn $ solve s

